src/lib/int.ts

```diff
@@ -0,0 +1,391 @@
+import { Bool, Circuit, Field } from '../snarky';
+import { CircuitValue, prop } from './circuit_value';
+
+function argToField(name: string, x: { value: Field } | number): Field {
+  if (typeof x === 'number') {
+    if (!Number.isInteger(x)) {
+      throw new Error(`${name} expected integer argument. Got ${x}`);
+    }
+    return new Field(x);
+  } else {
+    return x.value;
+  }
+}
+
+export class UInt64 extends CircuitValue {
+  @prop value: Field;
+
+  static get zero(): UInt64 {
+    return new UInt64(Field.zero);
+  }
+  constructor(value: Field) {
+    super();
+    this.value = value;
+  }
+
+  toString(): string {
+    return this.value.toString();
+  }
+
+  static check(x: UInt64) {
+    let actual = x.value.rangeCheckHelper(64);
+    actual.assertEquals(x.value);
+  }
+
+  static MAXINT(): UInt64 {
+    return new UInt64(Field.fromJSON(((1n << 64n) - 1n).toString()) as Field);
+  }
+
+  static fromNumber(x: number): UInt64 {
+    return new UInt64(argToField('UInt64.fromNumber', x));
+  }
+
+  static NUM_BITS = 64;
+
+  divMod(y: UInt64 | number): [UInt64, UInt64] {
+    let x = this.value;
+    let y_ = argToField('UInt64.div', y);
+
+    if (this.value.isConstant() && y_.isConstant()) {
+      let xn = BigInt(x.toString());
+      let yn = BigInt(y_.toString());
+      let q = xn / yn;
+      let r = xn - q * yn;
+      return [
+        new UInt64(new Field(q.toString())),
+        new UInt64(new Field(r.toString())),
+      ];
+    }
+
+    y_ = y_.seal();
+
+    let q = Circuit.witness(
+      Field,
+      () => new Field((BigInt(x.toString()) / BigInt(y_.toString())).toString())
+    );
+
+    q.rangeCheckHelper(UInt64.NUM_BITS).assertEquals(q);
+
+    // TODO: Could be a bit more efficient
+    let r = x.sub(q.mul(y_)).seal();
+    r.rangeCheckHelper(UInt64.NUM_BITS).assertEquals(r);
+
+    let r_ = new UInt64(r);
+    let q_ = new UInt64(q);
+
+    r_.assertLt(new UInt64(y_));
+
+    return [q_, r_];
+  }
+
+  /** Integer division.
```

Comment:
small nit: The comment style is unusual. Usually the `/**` line doesn't contain comment text https://typedoc.org/guides/doccomments/

Reply:
Addressed!